### ISSUE TITLE
Red index status causes NullPointerException in PySpark

### ISSUE THREAD

**Issue by jimmyjones2** [elastic/elasticsearch-hadoop#957]
Running spark-2.1.0-bin-hadoop2.7 with elasticsearch-hadoop-5.3.0.

Run up two elasticsearch nodes. Add documents and remove replicas:
```bash
curl -XPUT localhost:9200/test
curl -XPUT localhost:9200/test/test/1 -d '{"fish":"ing"}'
curl -XPUT localhost:9200/test/_settings -d '{"index":{"number_of_replicas":0}}'
curl -XPUT localhost:9200/test/test/2 -d '{"foo":"bar"}'
```

Kill one node, observe red status:
```bash
curl localhost:9200/_cat/health
1491835534 15:45:34 elasticsearch red 1 1 2 2 0 0 3 0 - 40.0%
```

Now run Spark query:
```bash
bin/pyspark --jars ../elasticsearch-hadoop-5.3.0/dist/elasearch-spark-20_2.11-5.3.0.jar
>>> f = sqlContext.read.format("org.elasticsearch.spark.sql").options(nodes="127.0.0.1:9200", nodes_wan_only="true").load("test")
>>> f.collect()
raceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jimmy/Downloads/spark-2.1.0-bin-hadoop2.7/python/pyspark/sql/dataframe.py", line 391, in collect
    port = self._jdf.collectToPython()
  File "/home/jimmy/Downloads/spark-2.1.0-bin-hadoop2.7/python/lib/py4j-0.10.4-src.zip/py4j/java_gateway.py", line 1133, in __call__
  File "/home/jimmy/Downloads/spark-2.1.0-bin-hadoop2.7/python/pyspark/sql/utils.py", line 63, in deco
    return f(*a, **kw)
  File "/home/jimmy/Downloads/spark-2.1.0-bin-hadoop2.7/python/lib/py4j-0.10.4-src.zip/py4j/protocol.py", line 319, in get_return_value
py4j.protocol.Py4JJavaError: An error occurred while calling o32.collectToPython.
: java.lang.NullPointerException
	at java.lang.StringBuilder.<init>(StringBuilder.java:112)
	at org.elasticsearch.hadoop.rest.RestService.findSlicePartitions(RestService.java:326)
	at org.elasticsearch.hadoop.rest.RestService.findPartitions(RestService.java:267)
	at org.elasticsearch.spark.rdd.AbstractEsRDD.esPartitions$lzycompute(AbstractEsRDD.scala:77)
	at org.elasticsearch.spark.rdd.AbstractEsRDD.esPartitions(AbstractEsRDD.scala:76)
	at org.elasticsearch.spark.rdd.AbstractEsRDD.getPartitions(AbstractEsRDD.scala:44)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:1958)
	at org.apache.spark.rdd.RDD$$anonfun$collect$1.apply(RDD.scala:935)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:362)
	at org.apache.spark.rdd.RDD.collect(RDD.scala:934)
	at org.apache.spark.sql.execution.SparkPlan.executeCollect(SparkPlan.scala:275)
	at org.apache.spark.sql.Dataset$$anonfun$collectToPython$1.apply$mcI$sp(Dataset.scala:2745)
	at org.apache.spark.sql.Dataset$$anonfun$collectToPython$1.apply(Dataset.scala:2742)
	at org.apache.spark.sql.Dataset$$anonfun$collectToPython$1.apply(Dataset.scala:2742)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:57)
	at org.apache.spark.sql.Dataset.withNewExecutionId(Dataset.scala:2765)
	at org.apache.spark.sql.Dataset.collectToPython(Dataset.scala:2742)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:280)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:214)
	at java.lang.Thread.run(Thread.java:745)
```
NB. in some cases, eg. if there were no documents I'd get a 503 propagated from Elasticsearch, but in this case when it can return some results get an NPE.

**Comment by jbaiera** [comment]
@jimmyjones2 I've got a clean reproduction of this. Thanks. I'll add some guarding statements in the code to keep it from throwing an NPE. In the case of a red cluster, I think it would make sense to throw at the start of the job, and ignore the red status if a given property is enabled, assuming users are ok with non-full results.

**Comment by jbaiera** [comment]
This is released in version 5.4.0
